### PR TITLE
Support invoke of function pointer type arg

### PR DESCRIPTION
**Pull request by steveharter**
Addresses the main issue of https://github.com/dotnet/runtime/issues/85028 which is the inability to invoke a method that has a function pointer type argument for CoreClr. Also adds support to Mono and fixes issue in NativeAot with returning function pointer types.

### CODE REVIEW

File: src/libraries/System.Reflection/tests/MethodInfoTests.cs [hunk 1]
```diff
@@ -894,8 +894,27 @@ private static void SecondCall(MethodInfo mi)
...
+            Assert.False((bool)m.Invoke(null, new object[] { (IntPtr)fn, 41 }));
+
+            // Verify return type; currently returned as a boxed IntPtr.
+            m = GetMethod(typeof(FunctionPointerMethods), nameof(FunctionPointerMethods.GetFunctionPointer));
+            object ret = m.Invoke(null, null);
+            Assert.IsType<IntPtr>(ret);
+            Assert.True((IntPtr)ret != 0);
+        }
+
+        // Mono will throw System.ArgumentException : Object of type 'System.IntPtr' cannot be converted to type 'System.Boolean(System.Int32)'.
```
Comment (jkotas) [inline src/libraries/System.Reflection/tests/MethodInfoTests.cs:915]:
```suggestion
```
This comment looks misplaced

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs [hunk 2]
```diff
@@ -3668,6 +3668,9 @@ private CheckValueStatus TryChangeTypeSpecial(ref object value)
             return CheckValueStatus.Success;
         }
 
+        // Remove this once Mono support for function pointers (#71095) is addressed.
+        internal bool IsIntPtrBasedFunctionPointer => IsFunctionPointer;
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:3672]: What are the failing tests without this workaround (outside the tests that are disabled against adding function pointer support to Mono)?

File: src/libraries/System.Reflection/tests/MethodInfoTests.cs [hunk 3]
```diff
@@ -894,8 +894,26 @@ private static void SecondCall(MethodInfo mi)
...
 
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/71095", TestRuntimes.Mono)]
+        private static unsafe void TestFunctionPointersWithFunctionPointerArgs()
+        {
+            void* fn = FunctionPointerMethods.GetFunctionPointer();
+            MethodInfo m;
+
+            m = GetMethod(typeof(FunctionPointerMethods), nameof(FunctionPointerMethods.CallFcnPtr_FP));
+            Assert.True((bool)m.Invoke(null, new object[] { (IntPtr)fn, 42 }));
+            Assert.False((bool)m.Invoke(null, new object[] { (IntPtr)fn, 41 }));
```
Comment (jkotas) [inline src/libraries/System.Reflection/tests/MethodInfoTests.cs:905]: Add test with UIntPtr as well? (I assume that both IntPtr and UIntPtr work - is that right?)

**Comment by ivanpovazan** [comment]
Is this planned to be merged in by RC1 snap? 
I am asking this because Mono support for function pointer introspection API is ready for review https://github.com/dotnet/runtime/pull/89712 and I would like to synchronise and agree what should get merged in first to avoid any regressions.

File: src/libraries/System.Reflection/tests/MethodInfoTests.cs [hunk 4]
```diff
@@ -894,8 +894,26 @@ private static void SecondCall(MethodInfo mi)
             Assert.Contains("TestAssembly", asm.ToString());
         }
 
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/71095", TestRuntimes.Mono)]
```
Comment (ivanpovazan) [inline src/libraries/System.Reflection/tests/MethodInfoTests.cs:897]: I think this method is missing `[Fact]` attribute.

**Comment by steveharter** [comment]
> Is this planned to be merged in by RC1 snap?

I'll hold off on this PR until the [Mono function pointer introspection PR](https://github.com/dotnet/runtime/pull/89712) is in.

**Comment by ivanpovazan** [comment]
> I'll hold off on this PR until the https://github.com/dotnet/runtime/pull/89712 is in.

Thank you. Done!

File: src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs (hunk 2 above)
Comment (jkotas) [reply src/coreclr/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs:3672]: I guess this workaround can be reverted now that the support for function pointers in Mono was merged.

File: src/libraries/System.Reflection/tests/MethodInfoTests.cs (hunk 3 above)
Comment (steveharter) [reply src/libraries/System.Reflection/tests/MethodInfoTests.cs]: Yep. Done

File: src/libraries/System.Reflection/tests/MethodInfoTests.cs (hunk 4 above)
Comment (steveharter) [reply src/libraries/System.Reflection/tests/MethodInfoTests.cs]: Thanks

**Comment by vargaz** [comment]
The mono changes look ok.

**Review by jkotas — APPROVED** [review]
LGTM. (Please check that the tests are passing on native AOT before merging - I have triggered extra-platforms run.)

File: src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/DynamicInvokeInfo.cs [hunk 5]
```diff
@@ -838,6 +838,11 @@ private unsafe object ReturnTransform(ref byte byref, bool wrapInTargetInvocatio
                 Debug.Assert(type.IsPointer);
                 obj = Pointer.Box((void*)Unsafe.As<byte, IntPtr>(ref byref), type);
             }
+            else if ((_returnTransform & Transform.FunctionPointer) != 0)
```
Comment (steveharter) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/System/Reflection/DynamicInvokeInfo.cs:841]: Without this change, a checked build would throw in the `else` statement below where it tried to box `_returnType`.

**Comment by steveharter** [comment]
> LGTM. (Please check that the tests are passing on native AOT before merging - I have triggered extra-platforms run.)

Also verified locally with both release and checked runtime build.

Did update https://github.com/dotnet/runtime/issues/90376 based on CI failures. Other failures appear unrelated.

**Pull request merged**